### ISSUE TITLE
Registering multiple products

### ISSUE THREAD

**Issue by leeDav** [j3k0/cordova-plugin-purchase#187]
Is there a knack or something I'm overlooking for registering many products?

For iOS I have:

``` javascript
store.register(
  [
    {
            id:    "com.lexisnexis.whillans.master",
            alias: "Master Unlock",
            type:  store.NON_CONSUMABLE
    },
    {
            id:    "com.lexisnexis.whillans.penaltiesandinterest",
            alias: "Penalties and Interest",
            type:  store.NON_CONSUMABLE
    }
]);
```

And this works fine. All products are registered. For Android however, I have exactly the same code and nothing works. I don't even get the error saying the app isn't signed/authorised for in app purchases.

Yet I remove all but one product, like so;

``` javascript
store.register(
    {
            id:    "com.lexisnexis.whillans.master",
            alias: "Master Unlock",
            type:  store.NON_CONSUMABLE
    });
```

And suddenly it works! What's going on? Am I doing it incorrectly?

**Comment by leeDav** [comment]
On further investigation I noticed the error ``getSkuDetails() failed: 5:Error`` in logcat. This led me to [this StackOverflow answer](http://stackoverflow.com/questions/20862280/googles-iab-api3-getskudetails-method-returns-failed-5-developer-error) which details the error. In short, Android doesn't -- or at least, didn't -- like it when you try to register more than 20 products at once.

The way around this, as detailed in the SO link, is to chunk the products in to smaller groups.

I copy/pasted the code from the answer in to my ``IabHelper.java`` file but this produced the error of ``IABHELPER_BAD_RESPONSE`` not being found. Reading the other answer led me to an updated version of the code provided by Google.

TL,DR: Update the Google library for IAP [as found here](https://code.google.com/p/marketbilling/source/browse/v3/src/com/example/android/trivialdrivesample/util/). This project though is deprecated in favour of the sample project which can be downloaded [through the ADK](https://developer.android.com/google/play/billing/billing_overview.html#samples).

I'd make a pull request but I'm very pushed for time, so if you see this before I get a chance to make one then great! If not, I'll try do one as soon as I can :)

**Comment by butlimous** [comment]
facing the same problem :(  Can not register multiple products in Android. Works in iOS

**Comment by chrisschaub** [comment]
Same issue. Should I just register one at a time then refresh? I'm also seeing an issue where, only on Android, if you re-refresh a product you get dupes in "store.products". iOS works fine.

**Comment by butlimous** [comment]
You should register one at a time and refresh only once. ex:
store.register ({ });
store.register ({ });
store.refresh();

It will work when you do it like this

**Comment by chrisschaub** [comment]
Ok, I'm doing that so I guess this is not my issue. Only on Android, I see multiple versions of product in store.products after a second refresh, all invalid. I'm going to try the git version now.

**Comment by leeDav** [comment]
I read somewhere that multiple calls to ```store.register()``` actually overwrite each other. That seemed to be true.

After replacing the ``IapHelper.java``` file manually I still encountered errors.

It was only until I removed the plugin and then installed it via the GitHub link (rather than the Cordova plugins site) that it correctly worked, as https://github.com/j3k0/cordova-plugin-purchase/issues/123#issuecomment-84715827 suggested.

(I'd ignore the stuff about editing the billing_config.xml manually)

In short (all one line);

```cordova plugin add https://github.com/j3k0/cordova-plugin-purchase.git -variable BILLING_KEY="<BILLING_KEY>```

**Comment by chrisschaub** [comment]
Thanks, yeah. It's almost working on android. I just can't get products to
become valid. iOS works fine.

1. My app is published in alpha
2. I'm a test user
3. The build is signed for release
4. The products register but remain invalid

I just can't get android to work. A total mystery since iOS is usually more
of a pain.


On Tue, May 26, 2015 at 4:02 PM, Lee Davies <notifications@github.com>
wrote:

> I read somewhere that multiple calls to store.register() actually
> overwrite each other. That seemed to be true.
>
> After replacing the IapHelper.java` file manually I still encountered
> errors.
>
> It was only until I removed the plugin and then installed it via the
> GitHub link (rather than the Cordova plugins site) that it correctly
> worked, as #123 (comment)
> <https://github.com/j3k0/cordova-plugin-purchase/issues/123#issuecomment-84715827>
> suggested.
>
> (I'd ignore the stuff about editing the billing_config.xml manually)
>
> In short (all one line);
>
> cordova plugin add https://github.com/j3k0/cordova-plugin-purchase.git
> -variable BILLING_KEY="<BILLING_KEY>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/j3k0/cordova-plugin-purchase/issues/187#issuecomment-105663955>
> .
>



-- 
Christopher Schaub
http://chris.schaub.com

**Comment by leeDav** [comment]
Are you able to post your code? Is the plugin's debug state set to verbose, so you get the most output (and can you share that?)

Some thoughts;
1. Have you waited several hours since uploading it as alpha? Some say 2-3 hours, for me, it was more like 8.
2. Are the products activated?

**Comment by chrisschaub** [comment]
Ok, I found my bug, and nothing to do with the plugin, my bad! I will say that on Android, multiple calls to register will populate store.products with dupes which is strange. But I just wrote around it. Doesn't happen on iOS. Thanks for all the help everybody!
